src/buildblock/ML_norm.cxx

```diff
@@ -518,758 +518,797 @@ float GeoData3D::operator()(const int ra, const int a, const int rb, const int b
+                proj_data_info_ptr->get_scanner_sptr()->
+                        get_num_axial_crystals_per_block();
+
+        const int new_num_transaxial_crystals_per_block = num_transaxial_crystals_per_block - virtual_transaxial_crystals;
+
+        const int new_num_axial_crystals_per_block = num_axial_crystals_per_block - virtual_axial_crystals;
+
+
+        const int num_transaxial_blocks_in_fansize = fan_size/(num_transaxial_crystals_per_block);
+        const int new_fan_size = fan_size - num_transaxial_blocks_in_fansize*virtual_transaxial_crystals;
+        const int new_half_fan_size = new_fan_size/2;
+        const int num_axial_blocks_in_max_delta = max_delta/(num_axial_crystals_per_block);
+        const int new_max_delta = max_delta - (num_axial_blocks_in_max_delta)*virtual_axial_crystals-1;
+        const int new_num_detectors_per_ring = num_detectors_per_ring - num_transaxial_blocks*virtual_transaxial_crystals;
+        const int new_num_rings = num_rings - (num_axial_blocks-1)*virtual_axial_crystals;
+//    std::cerr<<new_num_rings<<std::endl;
+//    std::cerr<<new_num_detectors_per_ring<<std::endl;
+//    std::cerr<<new_max_delta<<std::endl;
+//    std::cerr<<new_fan_size<<std::endl;
+
+
+        // ****    End     ****  //
+
+        fan_data = FanProjData(new_num_rings, new_num_detectors_per_ring, new_max_delta, 2*new_half_fan_size+1);
+
+
+        shared_ptr<SegmentBySinogram<float> > segment_ptr;
+        Bin bin;
+        for (bin.segment_num() = proj_data.get_min_segment_num(); bin.segment_num() <= proj_data.get_max_segment_num();  ++ bin.segment_num()) {
+            segment_ptr.reset(new SegmentBySinogram<float>(proj_data.get_segment_by_sinogram(bin.segment_num())));
+
+            for (bin.axial_pos_num() = proj_data.get_min_axial_pos_num(bin.segment_num());
+                 bin.axial_pos_num() <= proj_data.get_max_axial_pos_num(bin.segment_num());
+                 ++bin.axial_pos_num())
+                for (bin.view_num() = 0; bin.view_num() < num_detectors_per_ring / 2; bin.view_num()++)
+                    for (bin.tangential_pos_num() = -half_fan_size;
+                         bin.tangential_pos_num() <= half_fan_size;
+                         ++bin.tangential_pos_num()) {
+                        int ra = 0, a = 0;
+                        int rb = 0, b = 0;
+
+                        proj_data_info_ptr->get_det_pair_for_bin(a, ra, b, rb, bin);
+                        if ((ra == num_rings - 1) || (rb == num_rings - 1) || (a == num_detectors_per_ring - 1) ||
+                            (b == num_detectors_per_ring - 1)) {
+                            continue;
+                        }
+
+                        int offset = (a%num_transaxial_crystals_per_block)>new_num_transaxial_crystals_per_block?(a%num_transaxial_crystals_per_block%new_num_transaxial_crystals_per_block):0;
+                        int new_a = a - a/num_transaxial_crystals_per_block*virtual_transaxial_crystals - offset;
```

Comment:
```suggestion
                        int new_a = a - (a/num_transaxial_crystals_per_block)*virtual_transaxial_crystals;
```
I'd feel safer with brackets.

Reply:
`bool is_virtual_crystal(int a, int ra)
{
   const int a_in_block = a % num_transaxial_crystals_in_block;
   if (virtual_crystal_transaxial_after_block){
   if (a_in_block >= num_non_virtual_transaxial_crystals_in_block)
     return true;
   }
   else{
   if (a_in_block <= num_of_virtual_transaxial_crystals_in_block)
     return true;
}
const int ra_in_block = ra % num_axial_crystals_in_block;
if (virtual_crystal_axial_after_block){
   return (ra_in_block >= num_non_virtual_axial_crystals_in_block)
else
    return (ra_in_block <= num_of_virtual_axial_crystals_in_block)
}
}`

it is easy to add **virtual_crystal_transaxial_before_or_after_block**. And it can deal with outermost virtual rings in the axial.
Maybe it's not necessary to do such things now.
I think I can do some tests for mCT with and without virtual crystals.